docs/architecture/adr-009-revamp-testkit.md

```diff
@@ -0,0 +1,279 @@
+pub trait TestBlock: Clone + Debug {
+    /// The type of header can be extracted from the block.
+    type Header: TestHeader;
+
+    /// The height of the block.
+    fn height(&self) -> Height;
+
+    /// The timestamp of the block.
+    fn timestamp(&self) -> Timestamp;
+
+    /// Extract the header from the block.
+    fn into_header(self) -> Self::Header;
+}
+
+/// TestHeader is a trait that defines the interface for a header produced by a host blockchain.
+pub trait TestHeader: Clone + Debug + Into<Any> {
+    /// The type of consensus state can be extracted from the header.
+    type ConsensusState: ConsensusState + Into<AnyConsensusState> + From<Self> + Clone + Debug;
+
+    /// The height of the block, as recorded in the header.
+    fn height(&self) -> Height;
+
+    /// The timestamp of the block, as recorded in the header.
+    fn timestamp(&self) -> Timestamp;
+
+    /// Extract the consensus state from the header.
+    fn into_consensus_state(self) -> Self::ConsensusState;
+}
+```
+
+### Decoupling IbcContext and Host environment
+
+Currently, `MockContext` implements the top validation and execution context of
+`ibc-rs`. It contains other host-specific data e.g. `host_chain_id`,
+`block_time` - that are irrelevant to the IBC context directly. If we think
+`MockContext` as a real blockchain context, the `MockContext` represents the top
+runtime - which contains the IBC module. So we implement the validation and
+execution context on `MockIbcStore`, instead of `MockContext`.
+
+With this, the `MockContext` contains two decoupled parts - the host and the IBC
+module.
+
+### Chain-like interface for `MockContext`
+
+With the above changes, we can refactor the `MockContext` to have
+blockchain-like interfaces.
+
+The `MockContext` should have `end_block`, `produce_block` and `begin_block` to
+mimic the real blockchain environment such as Cosmos-SDK.
+
+```rs
+impl<S, H> MockContext<S, H>
+where
+    S: ProvableStore + Debug,
+    H: TestHost,
+{
+    pub fn ibc_store_mut(&mut self) -> &mut MockIbcStore<S>;
+    pub fn host_mut(&mut self) -> &mut H;
+
+    pub fn generate_genesis_block(&mut self, genesis_time: Timestamp);
+    pub fn begin_block(&mut self);
+    pub fn end_block(&mut self);
+    pub fn produce_block(&mut self, block_time: Duration);
+}
+```
+
+### ICS23 compatible proof generation
+
+With the new ability of proof generation, we can now test the Tendermint light
+clients. But we need our proofs to be ICS23 compatible. ICS23 expects the IBC
+store root to be committed at a commitment prefix at a top store in the host
+environment.
+
+For this, we add an extra store in `MockContext` where the `MockIbcStore`
+commits its storage root at its
+[commitment prefix](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/testapp/ibc/core/core_ctx.rs#L127-L129)
+key.
+
+So the `MockContext` is finalized as:
+
+```rs
+pub struct MockGenericContext<S, H>
```

Comment:
Is this `MockGenericContext` type the same as the above `MockContext` type, or is it a distinct type? If it's not distinct, let's make that more explicit, for example by using diff highlighting syntax:
```diff
-pub struct MockContext<S, H>
+pub struct MockGenericContext<S, H>
where
    S: ProvableStore + Debug,
    H: TestHost,
{
+   pub main_store: S,
+   pub host: H,
+   pub ibc_store: MockIbcStore<S>,

    pub fn ibc_store_mut(&mut self) -> &mut MockIbcStore<S>;
    pub fn host_mut(&mut self) -> &mut H;

    pub fn generate_genesis_block(&mut self, genesis_time: Timestamp);
    pub fn begin_block(&mut self);
    pub fn end_block(&mut self);
    pub fn produce_block(&mut self, block_time: Duration);
}
```

Reply:
You're right. We have it this way in the feature branch. But for the sake of this ADR, I wanted to keep it simple - as if we were just extending the `MockContext` on `v0.51.0`.

In the feature branch, `MockContext<H>` becomes `MockGenericContext<DefaultIbcStore, H>` to keep the tests simple.

Let me know what you feel. If you want to keep `MockGenericContext` - I will add the relevant explanation.

Reply:
Ok, then I think it would be clearer if we don't introduce the `MockGenericContext` name here, and just stick with `MockContext`; otherwise, it is a little confusing.